### PR TITLE
Fix crash when OGR layer is removed when source is stored

### PR DESCRIPTION
**Pull request by nyalldawson**
There's some bad behaviour in QgsOgrFeatureSource where the provider is stored and accessed in the feature source. This causes a crash if the layer is removed when a source has been created for it.

I.e. when removing a layer while a background map render is relying on it.

I've also added a test to the provider conformance suite to guard against this... let's see if any other providers have the same issue...

### CODE REVIEW

File: src/providers/ogr/qgsogrfeatureiterator.cpp [hunk 1]
```diff
@@ -353,7 +353,6 @@ bool QgsOgrFeatureIterator::readFeature( OGRFeatureH fet, QgsFeature &feature )
 
 
 QgsOgrFeatureSource::QgsOgrFeatureSource( const QgsOgrProvider *p )
-  : mProvider( p )
```
Comment (m-kuhn) [inline src/providers/ogr/qgsogrfeatureiterator.cpp]: 👍

File: src/providers/ogr/qgsogrfeatureiterator.h [hunk 2]
```diff
@@ -33,7 +33,6 @@ class QgsOgrFeatureSource : public QgsAbstractFeatureSource
     virtual QgsFeatureIterator getFeatures( const QgsFeatureRequest &request ) override;
 
   protected:
```
Comment (m-kuhn) [inline src/providers/ogr/qgsogrfeatureiterator.h]: Unrelated, just because it's visible in the diff: is ogrfeatureiterator intended to be subclassed?

Comment (nyalldawson) [inline src/providers/ogr/qgsogrfeatureiterator.h]: I very much doubt it. @wonder-sk?

**Comment by m-kuhn** [comment]
LGTM, the virtual layer iterator dependency on the provider really looks like a bit of trouble. Would it be safe to just use a `std::shared_ptr<sqlite3>` instead of a `QgsScopedSqlite` in QgsVirtualLayerProvider?

**Comment by nyalldawson** [comment]
> Would it be safe to just use a std::shared_ptr<sqlite3> instead of a QgsScopedSqlite in QgsVirtualLayerProvider?

Possibly. I thought the same, but honestly don't use virtual layers enough at the moment to be able to justify spending the time on a better fix here.

**Comment by m-kuhn** [comment]
Ok, merge it then, CC @mhugo

File: src/providers/ogr/qgsogrfeatureiterator.h (hunk 2 above)
Comment (wonder-sk) [inline src/providers/ogr/qgsogrfeatureiterator.h]: Nope, just a legacy from the times when we used to use "protected" by default :-)

Comment (nyalldawson) [inline src/providers/ogr/qgsogrfeatureiterator.h]: Ok, I'll fix

**Comment by nyalldawson** [comment]
@manisandro I've just noticed that the AFS provider has the provider tied very strongly into the iterator, so that will also suffer from this same crash.

**Comment by manisandro** [comment]
@nyalldawson Thanks for the headsup

**Pull request merged**